### PR TITLE
Better handling of scipy ode in integrator

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
Make our integrator re-entrant. Scipy'ode is can only have one active instance active for some of the methods. This cuased no issue when the whole integration is done at once as when used in `mesolve`. But with solver as class, this mean using them will have side effect on other solver instances.  This PR make the check for concurrent integrator usage before scipy does and force a reset if needed. Switching between system will be inefficient, but will return the expected results.

In lsoda's integrator's `_backstep` we are catching a warning before acting on it, but catching the warnings still printed them at the end of tests, bloating the output. In this PR, I also do the check before scipy does so the warning is never raised.

**Related issues or PRs**
Point raised in discussion with @hodgestar and @AGaliciaMartinez on #1710 

**Changelog**
Front run some safety check in scipy.ode

### CODE REVIEW

File: qutip/tests/solver/test_integrator.py [hunk 1]
```diff
@@ -101,3 +103,21 @@ def me_method(self, request):
...
     def mc_method(self, request):
         return request.param
+
+
+@pytest.mark.parametrize('integrator',
+    [IntegratorScipyZvode, IntegratorScipylsoda], ids=["zvode", "lsoda"])
+def test_concurent_usage(integrator):
+    sys = qutip.QobjEvo(qutip.qeye(1))
+    opt = SolverOdeOptions()
+    sys = qutip.QobjEvo(0.5*qutip.qeye(1))
```
Comment (AGaliciaMartinez) [inline qutip/tests/solver/test_integrator.py:113]: Is the first assign to `sys` necessary given that there is another one?

Comment (AGaliciaMartinez) [reply qutip/tests/solver/test_integrator.py]: I see it is now resolved. Thank you Eric!

File: qutip/solver/ode/scipy_integrator.py [hunk 2]
```diff
@@ -277,17 +304,16 @@ def _one_step(self, t):
         # Here we want to advance up to t doing maximum one step.
         # We check if a new step is needed.
         t_front = self._ode_solver._integrator.rwork[12]
+        safe_delta = self._ode_solver._integrator.rwork[11]/10 + 1e-15
```
Comment (hodgestar) [inline qutip/solver/ode/scipy_integrator.py:307]: Could you explain a little what the safe_delta change does and why it is better? I'm happy for it to stay in this PR but it seems unrelated to the concurrency checks?

File: qutip/solver/ode/scipy_integrator.py [hunk 3]
```diff
@@ -296,14 +322,13 @@ def _backstep(self, t):
...
-        # Scipy raise a warning, not an Error.
-        with warnings.catch_warnings():
-            warnings.filterwarnings("ignore")
+        t_front = self._ode_solver._integrator.rwork[12]
+        delta = self._ode_solver._integrator.rwork[11]
+        if t == self._ode_solver.t:
+            pass
+        elif t_front - delta <= t:
             self._ode_solver.integrate(t)
-        if not self._ode_solver.successful():
```
Comment (hodgestar) [inline qutip/solver/ode/scipy_integrator.py:306]: Do we not still need a `.successful()` check somewhere?

**Review by hodgestar — COMMENTED** [review]
I left some questions, but woot for finding a way to check for the concurrent usage.

**Review by hodgestar — APPROVED** [review]
